File: src/main/java/org/example/kalkulationsprogramm/controller/ZahlungsartController.java
package org.example.kalkulationsprogramm.controller;

import lombok.RequiredArgsConstructor;
import org.example.kalkulationsprogramm.config.FrontendUserPrincipal;
import org.example.kalkulationsprogramm.domain.Mitarbeiter;
import org.example.kalkulationsprogramm.domain.Zahlungsart;
import org.example.kalkulationsprogramm.repository.MitarbeiterRepository;
import org.example.kalkulationsprogramm.repository.ZahlungsartRepository;
import org.example.kalkulationsprogramm.service.BelegService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;

/**
 * REST-Endpoint fuer Zahlungsart-Stammdaten.
 *
 * Zugriff: dieselbe BELEG-Sehen-Berechtigung wie {@link SachkontoController}.
 * Anlegen/Editieren ist bewusst nicht ueber API moeglich — Stammdaten werden
 * per Flyway-Seed gepflegt; einzelne Eintraege koennen ueber DB-Migrationen
 * ergaenzt werden.
 */
@RestController
@RequestMapping("/api/buchhaltung")
@RequiredArgsConstructor
public class ZahlungsartController {

    private final ZahlungsartRepository zahlungsartRepository;
    private final BelegService belegService;
    private final MitarbeiterRepository mitarbeiterRepository;

    @GetMapping("/zahlungsarten")
    public ResponseEntity<List<Map<String, Object>>> list(
            @RequestParam(value = "nurAktive", defaultValue = "true") boolean nurAktive,
            @RequestParam(value = "token", required = false) String token,
            Authentication auth) {
        Mitarbeiter caller = resolveCaller(token, auth);
        if (caller == null || !belegService.darfSehen(caller)) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
        }
        List<Zahlungsart> zas = nurAktive
                ? zahlungsartRepository.findByAktivTrueOrderBySortierungAscBezeichnungAsc()
                : zahlungsartRepository.findAllByOrderBySortierungAscBezeichnungAsc();
        return ResponseEntity.ok(zas.stream().map(ZahlungsartController::toDto).toList());
    }

    private Mitarbeiter resolveCaller(String token, Authentication auth) {
        if (token != null && !token.isBlank()) {
            Mitarbeiter m = belegService.findByToken(token);
            if (m != null) return m;
        }
        if (auth != null && auth.getPrincipal() instanceof FrontendUserPrincipal p && p.getUsername() != null) {
            return mitarbeiterRepository.findAll().stream()
                    .filter(m -> p.getUsername().equalsIgnoreCase(m.getEmail()))
                    .findFirst().orElse(null);
        }
        return null;
    }

    private static Map<String, Object> toDto(Zahlungsart za) {
        return Map.of(
                "id", za.getId(),
                "bezeichnung", za.getBezeichnung(),
                "aktiv", za.isAktiv(),
                "sortierung", za.getSortierung()
        );
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/controller/ZahlungsartController.kt
package org.example.kalkulationsprogramm.controller

import org.example.kalkulationsprogramm.domain.Zahlungsart
import org.example.kalkulationsprogramm.repository.ZahlungsartRepository
import org.example.kalkulationsprogramm.service.BelegService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.Authentication
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/buchhaltung")
class ZahlungsartController(
    private val zahlungsartRepository: ZahlungsartRepository,
    private val belegService: BelegService
) {
    @GetMapping("/zahlungsarten")
    fun list(
        @RequestParam(value = "nurAktive", defaultValue = "true") nurAktive: Boolean,
        @RequestParam(value = "token", required = false) token: String?,
        auth: Authentication?
    ): ResponseEntity<List<Map<String, Any?>>> {
        val caller = belegService.findCaller(token, auth)
        if (caller == null || !belegService.darfSehen(caller)) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build()
        }
        val zahlungsarten = if (nurAktive) {
            zahlungsartRepository.findByAktivTrueOrderBySortierungAscBezeichnungAsc()
        } else {
            zahlungsartRepository.findAllByOrderBySortierungAscBezeichnungAsc()
        }
        return ResponseEntity.ok(zahlungsarten.map(::toDto))
    }

    companion object {
        private fun toDto(zahlungsart: Zahlungsart): Map<String, Any?> = mapOf(
            "id" to zahlungsart.id,
            "bezeichnung" to zahlungsart.bezeichnung,
            "aktiv" to zahlungsart.isAktiv(),
            "sortierung" to zahlungsart.sortierung
        )
    }
}
